inc/zoo/swar/SWAR.h

```diff
@@ -83,99 +88,173 @@ namespace detail {
+/// SIMD operations against that primitive type T treated as a SIMD register.
+/// SWAR operations are usually constant time, log(lane count) cost, or O(lane count) cost.
+/// Certain computaitonal workloads can be materially sped up using SWAR techniques.
+template<int NBits, typename T = uint64_t> struct SWAR {
     SWAR() = default;
     constexpr explicit SWAR(T v): m_v(v) {}
+    constexpr explicit operator T() const noexcept { return m_v; }
 
     constexpr T value() const noexcept { return m_v; }
 
-    constexpr SWAR operator|(SWAR o) { return SWAR(m_v | o.m_v); }
-    constexpr SWAR operator&(SWAR o) { return SWAR(m_v & o.m_v); }
-    constexpr SWAR operator^(SWAR o) { return SWAR(m_v ^ o.m_v); }
+    constexpr SWAR operator|(SWAR o) const noexcept { return SWAR(m_v | o.m_v); }
+    constexpr SWAR operator&(SWAR o) const noexcept { return SWAR(m_v & o.m_v); }
+    constexpr SWAR operator^(SWAR o) const noexcept { return SWAR(m_v ^ o.m_v); }
 
-    constexpr bool operator==(SWAR o) { return m_v == o.m_v; }
-
-    constexpr T at(int position) {
-        constexpr auto filter = (T(1) << Size) - 1;
-        return filter & (m_v >> (Size * position));
+    constexpr T at(int position) const noexcept {
+        constexpr auto filter = (T(1) << NBits) - 1;
+        return filter & (m_v >> (NBits * position));
     }
 
-    constexpr SWAR clear(int position) {
-        constexpr auto filter = (T(1) << Size) - 1;
-        auto invertedMask = filter << (Size * position);
+    constexpr SWAR clear(int position) const noexcept {
+        constexpr auto filter = (T(1) << NBits) - 1;
+        auto invertedMask = filter << (NBits * position);
         auto mask = ~invertedMask;
         return SWAR(m_v & mask);
     }
 
-    constexpr int top() { return msb_index(m_v) / Size; }
-    constexpr int fastIndex() { return __builtin_ctzll(m_v) / Size; }
+    /// The SWAR lane index that contains the MSB.  It is not the bit index of the MSB.
+    /// IE: 8 bit wide 32 bit SWAR: 0x0040'0000 will return 5, not 43.
+    constexpr int top() const noexcept { return msbIndex(m_v) / NBits; }
+    constexpr int lsbIndex() const noexcept { return __builtin_ctzll(m_v) / NBits; }
 
-    constexpr SWAR set(int index, int bit) {
-        return SWAR(m_v | (T(1) << (index * Size + bit)));
+    constexpr SWAR set(int index, int bit) const noexcept {
+        return SWAR(m_v | (T(1) << (index * NBits + bit)));
     }
 
-    constexpr operator bool() const { return m_v; }
-protected:
     T m_v;
 };
 
-template<int Size, typename T>
-struct BooleanSWAR: SWAR<Size, T> {
-    constexpr BooleanSWAR(T v): SWAR<Size, T>(v) {}
+/// Defining operator== on base SWAR types is entirely too error prone. Force a verbose invocation.
+template<int NBits, typename T = uint64_t> constexpr auto horizontalEquality(SWAR<NBits, T> left, SWAR<NBits, T> right) {
+  return left.m_v == right.m_v;
+}
+
+/// Isolating more than bits in type currently results in disaster.
```

Comment:
`NBits`

Reply:
Good catch.

Reply:
I do not understand why attempting to use ~ results in "can't shift negative numbers left" with unsigned integral types.